modin/pandas/base.py

```diff
@@ -2393,7 +2393,8 @@ def reset_index(
         # insert a dropped index into a DataFrame if these columns already
         # exist.
         if (
-            not drop
+            False  # have to avoid this check somehow
+            and not drop
             and not self._query_compiler.has_multiindex()
```

Comment:
As here `has_multiindex()` decides whether to access the index we may have a flag in the query compiler indicating if the index is MI. I bet most of the preceding functions are able to predict whether they produce MI or not and set an appropriate flag.